### PR TITLE
Commander datalinks lost failsafe handing

### PR DESCRIPTION
**Pull request by cmic0**
This PR takes over the original one #11454. 
In order to test the original PR some fixes have been added here.

**Tests**

- [X] data link loss & regain (GCS HEARTBEAT)
- [X] onboard computer link loss & regain (HEARTBEAT)
- [ ] iridium loss & regain -> **PENDING**

The above points have been tested in SITL.
The fixes performed in this PR prevent mavlink messages flooding when you lose ONBOARD Controller:
screenshot from 2019-02-26 11-29-11

After the fix:
screenshot from 2019-02-26 11-46-33

**NEXT STEPS**
- Unify datalink/OA status flags in vehicle_status_flags msg, e.g:
    * condition_data_link_valid
    * condition_avoidance_system_valid
    * condition_manual_control_valid
- Adapt  #11456 and #11457 to the new messages.

**REMARKS**
At the moment NAV_DLL_ACT and NAV_RCL_ACT are brought into `commander_params.c` but still has the old naming convention. 

@dagar @MaEtUgR @mrivi FYI

### CODE REVIEW

**Comment by MaEtUgR** [comment]
I now reviewed #11454, honstely it's a bit overwhelming how many changes are proposed at the same time. Part of it is probably duplicate because it's just what the other pr contains. Changing the base branch of the pr would probably help a lot.

File: src/modules/commander/Commander.cpp [hunk 1]
```diff
@@ -3884,203 +3825,183 @@ bool Commander::preflight_check(bool report)
...
-			if (!status.data_link_lost) {
-				mavlink_log_critical(&mavlink_log_pub, "ALL LOW LAT LINKS LOST, ACTIVATE HIGH LAT LINK");
+			if (_datalink_last_status_avoidance_system == telemetry_status_s::MAV_STATE_ACTIVE) {
+				mavlink_log_info(&mavlink_log_pub, "AVOIDANCE SYSTEM HEALTHY");
+			}
 
-			} else {
-				mavlink_log_critical(&mavlink_log_pub, "ACTIVATING AVAILABLE HIGH LATENCY LINK");
+			if (_datalink_last_status_avoidance_system == telemetry_status_s::MAV_STATE_CRITICAL) {
+				mavlink_log_info(&mavlink_log_pub, "AVOIDANCE SYSTEM TIMEOUT");
```
Comment (mrivi) [inline src/modules/commander/Commander.cpp]: I could prefer to keep this `mavlink_log_critical`

Comment (cmic0) [reply src/modules/commander/Commander.cpp]: I have had an offline talk with @baumanta about avoidance timeouts.. 
We came to the conclusion that a time out is less important than an avoidance loss, thus I opted for removing the warning (which I find annoying in QGC), and replaced with an info.

Comment (mrivi) [reply src/modules/commander/Commander.cpp]: the vehicle starts loitering when there is a timeout https://github.com/PX4/avoidance/blob/master/local_planner/src/nodes/local_planner_node_main.cpp#L75-L76

The pilot should be informed on why that's happening clearly. Since critical warning are also audible I would prefer to have it critical.

**Comment by cmic0** [comment]
> I would propose to try to bring in this PR as it is now and handle #11456 in a separate PR.
> As for #11457, it requires more discussion on the overall architecture so I would leave it for another PR.

Totally agree, I think that this should be the starting point as @baumanta commits are required to tests onboard controller lost/avoidance lost.

The vehicle_status_flag message modification can be handled in next PRs and then #11456 can be adapted on top of that.

**Comment by cmic0** [comment]
@MaEtUgR Major changes in this PR are the same as #11454 performed by @dagar.
Since to continue testing #11454 we needed @baumanta commits and other minor fixes we thought it could make sense to bring all those small fixes together and create new PR.

**Comment by cmic0** [comment]
Rebased on latest upstream.

**Comment by MaEtUgR** [comment]
> make sense to bring all those small fixes together and create new PR
ok, just as a general hint it's easier to have iterations when making big changes. #11454 is already quite big to review and it's clear that you need your adjustments for testing but try to avoid loading more and more changes in the pr you like to get merged since it will take longer to be reviewed, tested, rebased and it all gets slower in the end.

File: src/modules/commander/Commander.cpp (hunk 1 above)
Comment (baumanta) [reply src/modules/commander/Commander.cpp]: I think it is ok it leave it as an info. The timeout state is something very temporary and in most cases only perists over a fraction of a second. It point in time when a timeout is triggered is also somewhat random, as it can be tuned through a parameter in the avoidance repo. And if the sensor data is not recovered in a few seconds the planner will send abort anyway and you will see that as a critical message.

**Comment by thomasgubler** [comment]
@MaEtUgR we want to bring this in now. Do you have objections on the changes that claudio added on top of #11454 

@cmic0 I see some open review comments from @MaEtUgR  in #11454 ? Did you adress @MaEtUgR s reciew comments that he made in #11454 here? Please point that out to him by commenting to his comments with e.g. "fixed in [commit link]"

**Comment by cmic0** [comment]
@thomasgubler not yet, some of them are related on the general behavior (and not strictly related to the datalink failsafe handling) thus I preferred not to take actions.
I will start bringing in the quickest ones and discuss the others.

File: src/modules/commander/Commander.hpp [hunk 2]
```diff
@@ -155,27 +171,26 @@ class Commander : public ModuleBase<Commander>, public ModuleParams
...
-		uint64_t last_heartbeat = 0u;
-		uint64_t last_dl_loss = 0u;
-		bool preflight_checks_reported = false;
-		bool lost = true;
-		bool high_latency = false;
-	} _telemetry[ORB_MULTI_MAX_INSTANCES];
+	void		data_link_check(bool &status_changed);
+	int		_telemetry_status_sub[ORB_MULTI_MAX_INSTANCES] {};
+
+	uint64_t	_datalink_last_heartbeat_gcs{0};
```
Comment (bkueng) [inline src/modules/commander/Commander.hpp]: These should all be `hrt_abstime`

File: src/modules/commander/state_machine_helper.cpp [hunk 3]
```diff
@@ -834,48 +834,63 @@ void set_link_loss_nav_state(vehicle_status_s *status, actuator_armed_s *armed,
...
-
-	} else if (link_loss_act == link_loss_actions_t::AUTO_LAND && status_flags.condition_local_position_valid) {
-		status->nav_state = vehicle_status_s::NAVIGATION_STATE_AUTO_LAND;
-
-	} else if (link_loss_act == link_loss_actions_t::TERMINATE) {
-		status->nav_state = vehicle_status_s::NAVIGATION_STATE_TERMINATION;
-		armed->force_failsafe = true;
+	switch (link_loss_act) {
+	case (link_loss_actions_t::DISABLED):
+	case (link_loss_actions_t::AUTO_RECOVER):
```
Comment (bkueng) [inline src/modules/commander/state_machine_helper.cpp]:
```suggestion
	case link_loss_actions_t::AUTO_RECOVER:
```

File: src/modules/commander/Commander.cpp [hunk 4]
```diff
@@ -3888,203 +3829,183 @@ bool Commander::preflight_check(bool report)
...
+			telemetry_status_s telemetry;
 
-				preflight_check(true);
+			if (orb_copy(ORB_ID(telemetry_status), _telemetry_status_sub[i], &telemetry) == PX4_OK) {
 
-				// Provide feedback on mission state
-				const mission_result_s &mission_result = _mission_result_sub.get();
+				// handle different radio types
+				switch (telemetry.type) {
+				case (telemetry_status_s::LINK_TYPE_USB):
```
Comment (bkueng) [inline src/modules/commander/Commander.cpp]:
```suggestion
				case telemetry_status_s::LINK_TYPE_USB:
```

File: src/modules/commander/Commander.cpp [hunk 5]
```diff
@@ -3888,203 +3829,183 @@ bool Commander::preflight_check(bool report)
...
-				// if the high latency link is inactive we do not want to accidentally detect it as an active link
-				dl_loss_timeout_local = INT32_MIN;
-				dl_regain_timeout_local = INT32_MAX;
-			}
+					// Recover from data link lost
+					if (status.data_link_lost) {
+						if (hrt_elapsed_time(&_datalink_last_heartbeat_gcs) < (_datalink_regain_threshold.get() * 1_s)) {
+							status.data_link_lost = false;
+							status_changed = true;
+							mavlink_log_critical(&mavlink_log_pub, "DATA LINK REGAINED");
```
Comment (bkueng) [inline src/modules/commander/Commander.cpp]: Please go through all messages and avoid capitalizing them.
```suggestion
							mavlink_log_critical(&mavlink_log_pub, "Data link regained");
```

Also a regain does not have to be critical IMO.

File: src/modules/commander/Commander.cpp [hunk 6]
```diff
@@ -3888,203 +3829,183 @@ bool Commander::preflight_check(bool report)
...
 
-				_telemetry[i].lost = false;
-				have_link = true;
+						_datalink_last_heartbeat_avoidance_system = telemetry.heartbeat_time;
+						_datalink_last_status_avoidance_system = telemetry.remote_system_status;
 
-				if (!_telemetry[i].high_latency) {
-					have_low_latency_link = true;
-				}
+						if (_avoidance_system_lost != false) {
```
Comment (bkueng) [inline src/modules/commander/Commander.cpp]:
```suggestion
						if (_avoidance_system_lost) {
```
Also please change the type of `_avoidance_system_lost` to `bool`.

File: src/modules/commander/Commander.hpp [hunk 7]
```diff
@@ -155,27 +171,26 @@ class Commander : public ModuleBase<Commander>, public ModuleParams
...
+
+	uint64_t	_datalink_last_heartbeat_onboard_controller{0};
+	uint64_t	_onboard_controller_lost{0};
+
+	uint64_t	_datalink_last_heartbeat_avoidance_system{0};
+	uint64_t	_avoidance_system_lost{0};
+	uint64_t	_avoidance_system_not_started{0};
+	bool		_avoidance_system_status_change{0};
+	uint64_t	_datalink_last_status_avoidance_system{9};
```
Comment (bkueng) [inline src/modules/commander/Commander.hpp]:
```suggestion
	uint8_t	_datalink_last_status_avoidance_system{telemetry_status_s::MAV_STATE_UNINIT};
```

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 8]
```diff
@@ -1882,16 +1882,25 @@ MavlinkReceiver::handle_message_heartbeat(mavlink_message_t *msg)
 		mavlink_msg_heartbeat_decode(msg, &hb);
 
 		/* ignore own heartbeats, accept only heartbeats from GCS */
-		if (msg->sysid != mavlink_system.sysid && hb.type == MAV_TYPE_GCS) {
+		if (msg->sysid != mavlink_system.sysid || hb.type == MAV_TYPE_ONBOARD_CONTROLLER) {
```
Comment (bkueng) [inline src/modules/mavlink/mavlink_receiver.cpp]: Why did `MAV_TYPE_GCS` got removed? It means we'll accept heartbeats from other drones as telemetry status. I think we should limit to what we accept to our sysid or a GCS.
Something like:
```suggestion
		if ((msg->sysid != mavlink_system.sysid && hb.type == MAV_TYPE_GCS) || (msg->sysid == mavlink_system.sysid && hb.type == MAV_TYPE_ONBOARD_CONTROLLER)) {
```

Comment (cmic0) [reply src/modules/mavlink/mavlink_receiver.cpp]: your suggestion make sense to me. 
Done here 982e31587d80aa03441aeb6d6f4e6b2ceadae89e

File: src/modules/commander/Commander.cpp (hunk 5 above)
Comment (cmic0) [reply src/modules/commander/Commander.cpp]: For this suggestion and the ones below see here 4c37aa28d9453a0d05af08f8d8d239f98973adfd

File: boards/stm/32f4discovery/default.cmake [hunk 9]
```diff
@@ -0,0 +1,106 @@
+
```
[the remark this answers is not in the document]
Comment (cmic0) [reply boards/stm/32f4discovery/default.cmake]: Done here 0d6bb663e0868ba87d4a108f46766c3988afb482.

File: src/modules/commander/state_machine_helper.cpp [hunk 10]
```diff
@@ -834,48 +834,63 @@ void set_link_loss_nav_state(vehicle_status_s *status, actuator_armed_s *armed,
...
-		status->nav_state = vehicle_status_s::NAVIGATION_STATE_AUTO_RTL;
-
-	} else if (link_loss_act == link_loss_actions_t::AUTO_LAND && status_flags.condition_local_position_valid) {
-		status->nav_state = vehicle_status_s::NAVIGATION_STATE_AUTO_LAND;
-
-	} else if (link_loss_act == link_loss_actions_t::TERMINATE) {
-		status->nav_state = vehicle_status_s::NAVIGATION_STATE_TERMINATION;
-		armed->force_failsafe = true;
+	switch (link_loss_act) {
+	case (link_loss_actions_t::DISABLED):
```
[the remark this answers is not in the document]
Comment (cmic0) [reply src/modules/commander/state_machine_helper.cpp]: Done it here 5f9b82cbba23980f9ed5f40d5a82deb1048ae27f

File: src/modules/commander/Commander.hpp (hunk 2 above)
Comment (cmic0) [reply src/modules/commander/Commander.hpp]: Done here 1bb3b3fcfdc8bf375186b4a78c8e3052fceb933a

File: src/modules/commander/state_machine_helper.cpp (hunk 3 above)
Comment (cmic0) [reply src/modules/commander/state_machine_helper.cpp]: Done it here 5f9b82c

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 11]
```diff
@@ -1882,16 +1882,25 @@ MavlinkReceiver::handle_message_heartbeat(mavlink_message_t *msg)
...
-			tstatus.heartbeat_time = tstatus.timestamp;
-			tstatus.system_id = msg->sysid;
-			tstatus.component_id = msg->compid;
+			tstatus.heartbeat_time = hrt_absolute_time();
+			tstatus.remote_system_id = msg->sysid;
+			tstatus.remote_component_id = msg->compid;
+			tstatus.remote_type = hb.type;
+			tstatus.remote_system_status = hb.system_status;
+
+			if (_telem_status_pub == nullptr) {
```
[the remark this answers is not in the document]
Comment (cmic0) [reply src/modules/mavlink/mavlink_receiver.cpp]: Done here 483d60380e954907c6b13ba86ece4c09470c4284 and here 2241750d995f526071a14a9fa495b84e666e3503

File: msg/telemetry_status.msg [hunk 12]
```diff
@@ -1,18 +1,42 @@
...
 
-uint8 type				# type of the radio hardware
+uint8 remote_system_id			# system id of the remote system (Mavlink header sys_id)
+uint8 remote_component_id		# component id of the remote system (Mavlink header comp_id)
+uint8 remote_type			# remote system status flag (Mavlink MAV_TYPE)
+uint8 remote_system_status		# remote system status flag (Mavlink MAV_STATE)
 
-uint8 system_id				# system id of the remote system
-uint8 component_id			# component id of the remote system
+uint8 type				# type of the radio hardware
```
[the remark this answers is not in the document]
Comment (cmic0) [reply msg/telemetry_status.msg]: Done here 1ebcbd436c8fddab416440861f79441cf0fa4c29

**Comment by cmic0** [comment]
@bkueng I addressed all the points of your review; the code is ready to be merged.
Still haven't tested Iridiumbsd high latency telemetry by the way.

@MaEtUgR FYI: I managed to address some of your comments you made on #11454 in this commit 53d78c9aeae809b04d03e03ef0c1d240a3b7cfcb

File: src/modules/commander/Commander.hpp [hunk 13]
```diff
@@ -155,27 +169,28 @@ class Commander : public ModuleBase<Commander>, public ModuleParams
...
-		bool lost = true;
-		bool high_latency = false;
-	} _telemetry[ORB_MULTI_MAX_INSTANCES];
+	void		data_link_check(bool &status_changed);
+	int		_telemetry_status_sub{-1};
+
+	hrt_abstime	_datalink_last_heartbeat_gcs{0};
+
+	hrt_abstime	_datalink_last_heartbeat_onboard_controller{0};
+	bool 				_onboard_controller_lost{0};
```
[the remark this answers is not in the document]
Comment (cmic0) [reply src/modules/commander/Commander.hpp]: Ups! lost this comment.. done here 506425cec905360ec956524917f63284ba9b40ea.
